### PR TITLE
Compute grouping sets during planning

### PR DESCRIPTION
**Pull request by martint**
This change moves the expansion of grouping sets from the analyzer
to the planner. AST nodes for GroupingElement now only describe syntax.
The analyzer now only needs to process the original expressions in the
syntax tree (e.g., for the purpose of determining their type and validity)
instead of processing each instantiation of the expression in the expanded
grouping sets.

### CODE REVIEW

**Comment by martint** [comment]
This PR uncovered https://github.com/prestodb/presto/issues/11120, which requires more drastic changes to fix. Closing for now -- will reopen later with a fix for that issue, as well.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 1]
```diff
@@ -472,31 +478,35 @@ private PlanBuilder aggregate(PlanBuilder subPlan, QuerySpecification node)
...
+        List<Set<NodeRef<Expression>>> groupingSets = ImmutableList.of(ImmutableSet.of());
+        List<List<Symbol>> groupingSymbols = ImmutableList.of(ImmutableList.of());
+        if (node.getGroupBy().isPresent()) {
+            groupingSets = enumerateGroupingSets(node.getGroupBy().get().getGroupingElements());
+
+            Stream<List<Symbol>> stream = groupingSets.stream()
+                    .map(set -> set.stream()
+                            .map(reference -> groupingTranslations.get(reference.getNode()))
+                            .distinct()
+                            .collect(Collectors.toList()));
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: why not immutable list?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 2]
```diff
@@ -549,17 +559,96 @@ private PlanBuilder aggregate(PlanBuilder subPlan, QuerySpecification node)
...
+        }
+
+        return computeGroupingSetsCrossProduct(enumeratedSets);
+    }
+
+    private List<Set<NodeRef<Expression>>> computeGroupingSetsCrossProduct(List<List<Set<NodeRef<Expression>>>> enumeratedGroupingSets)
+    {
+        checkState(!enumeratedGroupingSets.isEmpty(), "enumeratedGroupingSets cannot be empty");
+
+        List<Set<NodeRef<Expression>>> crossProduct = new ArrayList<>();
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: why not immutable list builder?

File: presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java [hunk 3]
```diff
@@ -931,6 +932,18 @@ public void testGroupByEmpty()
         assertFails(MUST_BE_AGGREGATE_OR_GROUP_BY, "SELECT a FROM t1 GROUP BY ()");
     }
 
+    @Test
+    public void testComplexExpressionInGroupingSet()
+    {
+        assertFails(MUST_BE_COLUMN_REFERENCE, "\\Qline 1:49: GROUP BY expression must be a column reference: (x + 1)\\E", "SELECT 1 FROM (VALUES 1) t(x) GROUP BY ROLLUP(x + 1)");
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java]: can you please format this as:

```
assertFails(
    MUST_BE_COLUMN_REFERENCE, 
    "\\Qline 1:49: GROUP BY expression must be a column reference: (x + 1)\\E", 
    "SELECT 1 FROM (VALUES 1) t(x) GROUP BY ROLLUP(x + 1)");
```

File: presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4 [hunk 4]
```diff
@@ -163,22 +163,17 @@ groupBy
     ;
 
 groupingElement
```
Comment (kokosing) [inline presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4]: It looks like it should belong to different commit.  One thing is to move the grouping sets expensasion from analyzer to planner and other is to change how grouping sets expressions are parsed, validated and analyzed (before expansions). What do you think?
